Feature/component-recently-search

baraa0abd opened the pull request:
https://github.com/user-attachments/assets/b868a070-d7b8-4472-bd03-ff076fe2f5dd

baraa0abd committed:
feat: implement SearchTextInputField

This commit introduces a new composable function, `SearchTextInputField`, to the design system.

`SearchTextInputField` is a custom text input field specifically designed for search functionality. It includes:
- A leading search icon.
- A trailing clear icon that appears when text is entered.
- Placeholder hint text.
- Customizable colors for background, icons, hint text, and input text.
- Rounded corners for a modern look.

This component utilizes `BasicTextField` for core text input functionality and `MovioIcon` and `MovioText` for its visual elements.

baraa0abd committed:
chore: remove SearchTextInputField preview

This commit removes the preview composable for the `SearchTextInputField` component.

baraa0abd committed:
feat: implement RecentSearchHeader, RecentSearchItem, and RecentSearchList

This commit introduces three new composable components for displaying recent searches:

- `RecentSearchHeader`: Displays a header with the title "Recent Search" and a "Clear all" button.
- `RecentSearchItem`: Represents a single recent search query with a history icon, the search text, and a remove icon.
- `RecentSearchList`: A `Column` that displays a list of `RecentSearchItem` components.

baraa0abd committed:
feat: add string resources for RecentSearchHeader

This commit introduces string resources for "Recent Search" and "Clear all" text elements within the `RecentSearchHeader` composable.

The `RecentSearchHeader.kt` has been updated to use these string resources via `stringResource(id = R.string.resource_name)`, enhancing localization capabilities.

baraa0abd committed:
feat: implement RecentSearchScreen

This commit introduces a new composable function, `RecentSearchScreen`, which displays a search input field and a list of recent searches.

Key features:
- Uses `SearchTextInputField` for search input.
- Displays a `RecentSearchHeader`.
- Populates a `RecentSearchList` with sample recent search terms.
- Styled with a white background and rounded corners.

baraa0abd committed:
feat: add WorkManager and Koin work manager dependencies

This commit adds the following dependencies to the `presentation` module:
- `androidx.work:work-runtime-ktx:2.9.0` for using Android WorkManager.
- `io.insert-koin:koin-androidx-workmanager:3.5.3` for Koin integration with WorkManager.

It also includes a duplicate `io.insert-koin:koin-android:3.5.3` dependency.

baraa0abd committed:
Refactor: Migrate search UI to DesignSystem components and update ViewModel

This commit refactors the search UI by:

1.  **Deleting `RecentSearchList.kt` and `SearchHeader.kt`**: These custom composables have been replaced with new components from the `DesignSystem` module.
2.  **Using DesignSystem components**:
    *   `SearchHeader` is replaced by `SearchTextInputField` and `RecentSearchHeader` from the DesignSystem.
    *   `RecentSearchList` is replaced by the `RecentSearchList` component from the DesignSystem.
3.  **Updating `SearchListViewModel`**:
    *   Now uses `RecentSearchRepository` interface instead of `InMemoryRecentSearchRepository` directly.
    *   ViewModel methods now launch coroutines using `viewModelScope`.
    *   `removeRecentSearch` now takes a `RecentSearchItem` instead of an ID string.
    *   `clearAll` is renamed to `clearAllRecentSearches`.
4.  **Updating `SearchScreen`**:
    *   Integrates the new DesignSystem components for search input, recent search header, and recent search list.
    *   Adjusts logic for adding and removing recent searches to align with the updated ViewModel.
    *   Applies appropriate styling and layout using `AppTheme`.

baraa0abd committed:
Merge remote-tracking branch 'origin/develop' into feature/component-recently-searched

baraa0abd committed:
Refactor: Enable clearing of recent searches

The `RecentSearchHeader` component's "Clear All" text is now clickable, triggering the `onClearAll` lambda when pressed.

The explicit height and width modifiers on the "Clear All" text have been removed to allow the clickable area to encompass the text more naturally.

baraa0abd committed:
Refactor: Add click handling to RecentSearchList and keyboard actions to SearchTextInputField

This commit introduces two key enhancements:

1.  **Clickable RecentSearchList items:**
    *   Adds an `onItemClick` callback to `RecentSearchList` to handle clicks on individual recent search items.
    *   Updates `RecentSearchItem` to be clickable and trigger the new `onClick` callback.

2.  **Keyboard actions for SearchTextInputField:**
    *   Adds `imeAction` and `onImeAction` parameters to `SearchTextInputField`.
    *   Configures `KeyboardOptions` and `KeyboardActions` to trigger `onImeAction` on "Done" or "Search" keyboard actions.

baraa0abd committed:
Refactor: Update SearchListViewModel to use RecentSearchRepository

This commit refactors the `SearchListViewModel` to utilize the `RecentSearchRepository` for managing recent search data.

- The constructor of `SearchListViewModel` now expects an instance of `RecentSearchRepository`.
- Methods for loading, adding, removing, and clearing recent searches (`loadRecentSearches`, `addRecentSearch`, `removeRecentSearch`, `clearAllRecentSearches`) now use `viewModelScope.launch` to interact with the repository and update the `_recentSearches` StateFlow.
- The `removeRecentSearch` method now takes a `RecentSearchItem` as a parameter and uses its ID.
- The `clearAll` method has been renamed to `clearAllRecentSearches`.

Additionally, unused `androidx.compose.ui.tooling.preview.Preview` imports were removed from the following design system components:
- `RecentSearchHeader.kt`
- `RecentSearchItem.kt`
- `RecentSearchList.kt`
- `SearchTextInputField.kt`

The Koin module in `app.kt` has been updated to reflect the change in `SearchListViewModel`'s package and constructor.

baraa0abd committed:
Refactor: Move SearchListViewModel to presentation layer and use coroutines

This commit moves the `SearchListViewModel` from the `app` module's `screens` package to the `presentation` module's `component.screens.SearchScreen` package.

Additionally, the `SearchListViewModel` has been refactored to:
- Depend on the `RecentSearchRepository` interface from the `domain` layer instead of the concrete `InMemoryRecentSearchRepository`.
- Utilize `viewModelScope.launch` for asynchronous operations when loading, adding, removing, and clearing recent searches.
- Update the `removeRecentSearch` method to accept a `RecentSearchItem` object and extract the ID from it.

The following file was moved:
- `app/src/main/java/com/example/movio/screens/SearchListViewModel.kt` -> `presentation/src/main/java/com/example/presentation/component/screens/SearchScreen/SearchListViewModel.kt`

baraa0abd committed:
Refactor: Enhance recent search functionality and UI

This commit introduces several improvements to the recent search feature:

- **UI Highlighting:** `RecentSearchItem` and `RecentSearchList` now accept a `searchQuery` parameter. When a recent search item starts with the current `searchQuery`, the matching part of the text is highlighted.
- **ViewModel Refactoring:** `SearchListViewModel` has been updated to use `RecentSearchRepository` (from the domain layer) instead of `InMemoryRecentSearchRepository`. Methods now operate within `viewModelScope.launch` for better asynchronous handling.
- **IME Action for Adding Search:** In `searchListScreen`, pressing the IME action (e.g., "Search" on the keyboard) now adds the current `searchQuery` to the recent searches if it's not blank.
- **Passing Search Query to List:** `searchListScreen` now passes the current `searchQuery` to the `RecentSearchList` composable for highlighting.

These changes improve the user experience by visually connecting the current search with recent search suggestions and streamline the underlying data handling in the ViewModel.

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchHeader.kt:
```diff
@@ -0,0 +1,44 @@
+package com.example.designsystem.component.search
+
```
this should not be in the designsystem
it should be in the presentation

baraa0abd replied:
it's not a screen, it's a component, that is what I thought

fatimafaisalghazi replied:
yeah but it is a designsystem component it is a shared component in the presentation

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchHeader.kt:
```diff
@@ -0,0 +1,44 @@
+        modifier = modifier
+            .fillMaxWidth()
+            .height(19.dp)
+            .width(360.dp),
```
remove width(360.dp) you are already have fillMaxWidth

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchHeader.kt:
```diff
@@ -0,0 +1,44 @@
+            text = stringResource(id = R.string.clear_all),
+            textStyle = AppTheme.textStyle.label.smallRegular14,
+            color = AppTheme.colors.surfaceColor.onSurfaceVariant,
+            modifier = Modifier.clickable { onClearAll() }
```
you can add clip here to modify the click ripple

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+package com.example.designsystem.component.search
```
this too should be in the presentaion

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+import com.example.designsystem.component.MovioIcon
+import com.example.designsystem.component.MovioText
+import androidx.compose.foundation.layout.Column
+import androidx.compose.material3.Text
```
dont use material3

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+    Row(
+        modifier = modifier
+            .height(40.dp)
+            .width(328.dp)
```
use fillmaxwidth with padding horizotal 16

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+        Spacer(modifier = Modifier.width(8.dp))
+        Box(modifier = Modifier.weight(1f)) {
+            if (searchQuery.isNotBlank() && text.startsWith(searchQuery, ignoreCase = true)) {
+                Text(
+                    buildAnnotatedString {
+                        withStyle(style = androidx.compose.ui.text.SpanStyle(color = AppTheme.colors.surfaceColor.onSurface)) {
+                            append(text.substring(0, searchQuery.length))
+                        }
+                        withStyle(style = androidx.compose.ui.text.SpanStyle(color = AppTheme.colors.surfaceColor.onSurfaceVariant)) {
+                            append(text.substring(searchQuery.length))
+                        }
+                    },
+                    style = AppTheme.textStyle.label.smallRegular14
```
modify MovioText and add buildAnnotatedString  as optional parameter

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+            }
+        }
+        MovioIcon(
+            painter = painterResource(id = R.drawable.outline_add), // X icon
```
delete this comment

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt:
```diff
@@ -0,0 +1,76 @@
+        }
+        MovioIcon(
+            painter = painterResource(id = R.drawable.outline_add), // X icon
+            contentDescription = "Remove Icon",
```
extract it as string resource

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchList.kt:
```diff
@@ -0,0 +1,25 @@
+package com.example.designsystem.component.search
```
the same

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/textInputField/SearchTextInputField.kt:
```diff
@@ -0,0 +1,99 @@
+        cursorBrush = SolidColor(iconColor),
+        modifier = modifier
+            .height(48.dp)
+            .width(360.dp)
```
use fillmaxwidth

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/textInputField/SearchTextInputField.kt:
```diff
@@ -0,0 +1,99 @@
+        modifier = modifier
+            .height(48.dp)
+            .width(360.dp)
+            .background(backgroundColor, RoundedCornerShape(16.dp)),
```
use designsystem radius

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/textInputField/SearchTextInputField.kt:
```diff
@@ -0,0 +1,99 @@
+                verticalAlignment = Alignment.CenterVertically,
+                modifier = Modifier
+                    .fillMaxSize()
+                    .padding(horizontal = 16.dp)
```
use designsystem spacing

fatimafaisalghazi commented on designSystem/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,5 @@
+<resources>
+    <string name="recent_search">Recent Search</string>
+    <string name="clear_all">Clear all</string>
+</resources> 
```
locolaze this to arabic

fatimafaisalghazi commented on presentation/build.gradle.kts:
```diff
@@ -64,5 +64,8 @@ dependencies {
 
     implementation("androidx.lifecycle:lifecycle-viewmodel-ktx:2.9.1")
+    implementation("androidx.work:work-runtime-ktx:2.9.0")
+    implementation("io.insert-koin:koin-android:3.5.3")
```
this is a duplicated dependencies

fatimafaisalghazi commented on presentation/src/main/java/com/example/presentation/component/searchListUi/searchListScreen.kt:
```diff
@@ -36,42 +40,48 @@ fun SearchScreen() {
+            items = filteredRecentSearches.map { it.query },
+            modifier = Modifier
+                .align(Alignment.CenterHorizontally)
+                .width(360.dp),
```
the same here

fatimafaisalghazi commented on presentation/src/main/java/com/example/presentation/component/searchListUi/searchListScreen.kt:
```diff
@@ -36,42 +40,48 @@ fun SearchScreen() {
+            },
+            modifier = Modifier
+                .align(Alignment.CenterHorizontally)
+                .width(360.dp)
```
fillmaxwidth

fatimafaisalghazi commented on presentation/src/main/java/com/example/presentation/component/searchListUi/searchListScreen.kt:
```diff
@@ -36,42 +40,48 @@ fun SearchScreen() {
+        RecentSearchHeader(
+            modifier = Modifier
+                .align(Alignment.CenterHorizontally)
+                .width(360.dp),
```
the same here

fatimafaisalghazi reviewed (changes requested):
good work please resolve the comments

baraa0abd committed:
Refactor: Update SearchListViewModel to use RecentSearchRepository and move search components

This commit refactors the `SearchListViewModel` to utilize the `RecentSearchRepository` interface from the domain layer and updates related UI components.

- `SearchListViewModel` now depends on `RecentSearchRepository` and uses `viewModelScope.launch` for asynchronous operations when loading, adding, removing, and clearing recent searches. The `removeRecentSearch` method now accepts a `RecentSearchItem` and extracts the ID.
- `RecentSearchHeader`, `RecentSearchItem`, and `RecentSearchList` components have been moved from the `designSystem` module to the `presentation` module under `component.search`.
- `RecentSearchScreen` now uses `SearchListViewModel` (with `InMemoryRecentSearchRepository`) to manage and display recent searches.
- `MovioText` in `designSystem` now supports `AnnotatedString` for more flexible text styling.
- `SearchTextInputField` in `designSystem` has been updated to use `fillMaxWidth` and predefined spacing/radius values.
- Unnecessary Koin Android dependency was removed from `presentation/build.gradle.kts`.

The following files were moved:
- `designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchHeader.kt` -> `presentation/src/main/java/com/example/presentation/component/search/RecentSearchHeader.kt` (Note: content was deleted from the original and similar content now exists in the new location, effectively a move with modifications)
- `designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchItem.kt` -> `presentation/src/main/java/com/example/presentation/component/search/RecentSearchItem.kt` (Note: content was deleted from the original and similar content now exists in the new location, effectively a move with modifications)
- `designSystem/src/main/java/com/example/designsystem/component/search/RecentSearchList.kt` -> `presentation/src/main/java/com/example/presentation/component/search/RecentSearchList.kt` (Note: content was deleted from the original and similar content now exists in the new location, effectively a move with modifications)

baraa0abd committed:
Refactor: Introduce new `search` package for recent search UI components and update `SearchListViewModel`

This commit introduces a new `search` package under `presentation/src/main/java/com/example/presentation/component/` and moves/creates UI components related to recent searches into it.

**New UI Components in `presentation.component.search`:**
- `RecentSearchHeader.kt`: A composable for displaying the "Recent search" title and a "Clear all" button.
- `RecentSearchItem.kt`: A composable for displaying an individual recent search item with an icon, text (with potential highlighting of the `searchQuery`), and a remove icon.
- `RecentSearchList.kt`: A composable for displaying a list of `RecentSearchItem`s.

**`SearchTextInputField.kt` changes:**
- Modified `Spacer` width from `fillMaxWidth` to `10.dp` to better accommodate the layout within the text field.

**`SearchListViewModel.kt` changes:**
- Now depends on the `RecentSearchRepository` interface from the `domain` layer instead of the concrete `InMemoryRecentSearchRepository`.
- Utilizes `viewModelScope.launch` for asynchronous operations when loading, adding, removing, and clearing recent searches.
- Updates the `removeRecentSearch` method to accept a `RecentSearchItem` object and extract the ID from it.
- Renamed `clearAll()` to `clearAllRecentSearches()`.

The following files were created:
- `presentation/src/main/java/com/example/presentation/component/search/RecentSearchHeader.kt`
- `presentation/src/main/java/com/example/presentation/component/search/RecentSearchItem.kt`
- `presentation/src/main/java/com/example/presentation/component/search/RecentSearchList.kt`

baraa0abd committed:
Refactor: Update SearchListViewModel to use RecentSearchRepository

This commit refactors the `SearchListViewModel` to utilize the `RecentSearchRepository` for managing recent search data.

- The constructor of `SearchListViewModel` now expects an instance of `RecentSearchRepository`.
- Methods for loading, adding, removing, and clearing recent searches (`loadRecentSearches`, `addRecentSearch`, `removeRecentSearch`, `clearAllRecentSearches`) now use `viewModelScope.launch` to interact with the repository and update the `_recentSearches` StateFlow.
- The `removeRecentSearch` method now takes a `RecentSearchItem` as a parameter and uses its ID.
- The `clearAll` method has been renamed to `clearAllRecentSearches`.

Additionally, a new Arabic translation file (`values-ar/strings.xml`) has been added to the `designSystem` module, providing translations for "Recent Search", "Clear All", "Remove", and "History Icon".

baraa0abd committed:
Refactor: Standardize component widths and theme usage

This commit refactors several UI components to use `fillMaxWidth` instead of fixed widths, ensuring they adapt to different screen sizes. It also updates components to consistently use `AppTheme` for styling, specifically for colors, text styles, and spacing.

Key changes:

- **Width Adjustment:**
    - `SearchScreen`: `SearchTextInputField`, `RecentSearchHeader`, and `RecentSearchList` now use `fillMaxWidth()` instead of a fixed `width(360.dp)`.
- **Theme Consistency:**
    - `RecentSearchHeader`: Now uses `AppTheme.radius.medium` for corner rounding.
    - `SearchTextInputField`: Now uses `AppTheme.spacing.medium` for horizontal padding and `AppTheme.radius.medium` for corner rounding.
- **Content Description:**
    - `RecentSearchItem`: Updated content descriptions for "History Icon" and "Remove Icon" to be more specific.
- **Import Adjustments:**
    - `RecentSearchList.kt`: Package path updated from `com.example.presentation.component.search` to `com.example.presentation.component`.
    - `RecentSearchScreen.kt`: The `items` parameter for `RecentSearchList` now correctly maps `RecentSearchItem` to `item.query`.
    - Minor import changes across files to reflect the movement of `RecentSearchList` and to remove unused imports.

The `SearchListViewModel.kt` files in both `presentation/src/main/java/com/example/presentation/component/screens/searchscreen/` and `presentation/src/main/java/com/example/presentation/component/searchListUi/` were updated to import `RecentSearchItem` from `com.example.domain` instead of `com.example.presentation.component.search`.

baraa0abd committed:
Refactor: Update recent search functionality and add string resource

This commit introduces the following changes:

- **Recent Search Removal:** In `searchListScreen.kt`, the `removeRecentSearch` function in the ViewModel is now called with the `id` of the `RecentSearchItem` instead of its `query`.
- **Recent Search Item Click:** When a recent search item is clicked in `searchListScreen.kt`, the `searchQuery` is now updated with the `query` of the clicked item.
- **`RecentSearchItem` Import:** The import for `RecentSearchItem` in `SearchListViewModel.kt` was updated to `com.example.domain.RecentSearchItem`.
- **String Resource:** A new string resource `remove_icon` with the value "Remove Icon" (and "إزالة" for Arabic) has been added to `strings.xml`. The Arabic translation for `remove_icon_desc` was also updated to "إزالة".

ahmedhgabr commented on presentation/src/main/java/com/example/presentation/component/screens/searchscreen/SearchListViewModel.kt:
```diff
@@ -1,45 +1,33 @@
-
+package com.example.presentation.component.screens.SearchScreen
 
+import com.example.data.InMemoryRecentSearchRepository
```
Presentation/view model shouldn't depend on data. Check with the domain team and change it to `UseCase`

fatimafaisalghazi commented on designSystem/src/main/java/com/example/designsystem/component/MovioText.kt:
```diff
@@ -6,19 +6,33 @@ import androidx.compose.ui.Modifier
 @Composable
 fun MovioText(
-    text: String,
+    text: String = "",
```
it must not be an optional parameter

baraa0abd replied:
actually, I need it in the search to be optional
